radicle-link-test/src/daemon/common.rs

```diff
@@ -0,0 +1,256 @@
+    expected_remote: PeerId,
+) -> Result<(), anyhow::Error> {
+    assert_event!(
+        receiver,
+        PeerEvent::RequestCloned(urn, remote_peer) if urn == *expected_urn && remote_peer == expected_remote
+    ).context("assert_cloned")
+}
+
+pub async fn assert_fetched(
+    receiver: &mut broadcast::Receiver<PeerEvent>,
+) -> Result<(), anyhow::Error> {
+    assert_event!(receiver, PeerEvent::GossipFetched { .. }).context("assert_fetched")
+}
+
+/// Assert that we received a query event for the expected `RadUrn`.
+pub async fn requested(
+    receiver: &mut broadcast::Receiver<PeerEvent>,
+    expected: &Urn,
+) -> Result<(), anyhow::Error> {
+    assert_event!(
+        receiver,
+        PeerEvent::RequestQueried(urn) if urn == *expected
+    )
+    .context("requested")
+}
+
+/// Assert that the `PeerStatus` transitions to `Online` and the number of
+/// connected peers is equal to or more than `min_connected`.
+async fn connected(
+    receiver: &mut broadcast::Receiver<PeerEvent>,
+    min_connected: usize,
+) -> Result<(), anyhow::Error> {
+    assert_event!(
+        receiver,
+        PeerEvent::StatusChanged {
+            new: PeerStatus::Online { connected_peers, .. },
+            ..
+        } if connected_peers.len() >= min_connected
+    )
+    .context("connected")
+}
+
+async fn started(receiver: &mut broadcast::Receiver<PeerEvent>) -> Result<(), anyhow::Error> {
+    assert_event!(
+        receiver,
+        PeerEvent::StatusChanged {
+            new: PeerStatus::Started,
+            ..
+        }
+    )
+    .context("started")
+}
+
+pub struct PeerHandle {
+    pub peer_id: PeerId,
+    pub listen_addrs: Vec<SocketAddr>,
+    pub path: PathBuf,
+    pub owner: LocalIdentity,
+    pub peer: librad::net::peer::Peer<BoxedSigner>,
+    pub events: broadcast::Receiver<PeerEvent>,
+    pub control: peer::Control,
+}
+
+pub struct Harness {
+    tasks: Vec<tokio::task::JoinHandle<()>>,
+    rt: tokio::runtime::Runtime,
+    tmp: Vec<tempfile::TempDir>,
+}
+
+impl Harness {
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            tasks: Vec::new(),
+            rt: tokio::runtime::Builder::new_multi_thread()
+                .enable_all()
+                .build()
+                .unwrap(),
+            tmp: Vec::new(),
+        }
+    }
+
+    pub fn add_peer<S: AsRef<str>>(
+        &mut self,
+        owner_name: S,
+        run_config: RunConfig,
+        seeds: &[Seed],
+    ) -> Result<PeerHandle, anyhow::Error> {
+        let tmp = tempfile::tempdir()?;
+        let key = SecretKey::new();
+        let signer = signer::BoxedSigner::from(key);
+        let store = kv::Store::new(kv::Config::new(tmp.path().join("store")))?;
+        let paths = Paths::from_root(tmp.path())?;
+        let conf = config::configure(paths, signer, *config::LOCALHOST_ANY);
+        let disco = config::static_seed_discovery(seeds);
+        let peer = {
+            let _enter = self.rt.enter();
```

Comment:
This one is a bit unfortunate. To get the runtime handle for `librad::executor::Spawner`, we either need to pass it down to the library, or make otherwise sync code run inside some runtime context. Not sure how to go about that.